### ISSUE TITLE
static caching in MembershipManager::getMemberships() doesn't work properly with states

### ISSUE THREAD

**Issue by joachim-n** [Gizra/og#336]
```
  public function getMemberships(AccountInterface $user, array $states = [OgMembershipInterface::STATE_ACTIVE]) {
    // Get a string identifier of the states, so we can retrieve it from cache.
    sort($states);
    $states_identifier = implode('|', array_unique($states));

    $identifier = [
      __METHOD__,
      'user',
      $user->id(),
      $states_identifier,
    ];
    $identifier = implode(':', $identifier);

    // Return cached result if it exists.
    if (isset($this->cache[$identifier])) {
      return $this->cache[$identifier];
    }
```

If I ask for the memberships for user X, with all states, I pass in an empty array for $states (though that's not documented - #335). The cache ID is then:

> "Drupal\og\MembershipManager::getMemberships:user:1:"

If I then ask for the memberships for user X which are active, the cache ID is:

> Drupal\og\MembershipManager::getMemberships:user:1:active"

That results in a cache miss, even though the data is loaded already.

**Comment by pfrenssen** [comment]
Nice find. Together with the documentation issue in #335 I think we should introduce a new array constant that contains all possible states and document that this should be used.

We can keep the empty array passing too. We have a bunch of tests that verify that this works so this was clearly intended as such, but I think it would be better to just document that people should pass `OgMembershipInterface::STATES` since this will make it clearer what is going to be returned.

**Comment by amitaibu** [comment]
> but I think it would be better to just document that people should pass OgMembershipInterface::STATES since this will make it clearer what is going to be returned.

I agree

**Comment by pfrenssen** [comment]
This has been fixed in #449
